src/ts/miner.ts

```diff
@@ -0,0 +1,153 @@
+import assert from "assert";
+
+import {
+  BigNumber,
+  BytesLike,
+  Contract,
+  ContractFactory,
+  Signer,
+  ethers,
+} from "ethers";
+
+const OPCODES: Record<string, number | undefined> = {
+  STOP: 0x00,
+  ISZERO: 0x15,
+  ADDRESS: 0x30,
+  CALLVALUE: 0x34,
+  RETURNDATASIZE: 0x3d,
+  COINBASE: 0x41,
+  MSTORE: 0x52,
+  JUMPI: 0x57,
+  GAS: 0x5a,
+  CALL: 0xf1,
+  RETURN: 0xf3,
+};
+const PUSH1_OPCODE = 0x60;
+
+function assemble(source: string): BytesLike {
+  const bytecode = source
+    .split("\n")
+    .map((line) => line.replace(/;;.*$/, "").trim())
+    .filter((line) => line.length > 0)
+    .map((line) => {
+      const [op, param] = line.split(" ").map((part) => part.trim());
+      // op-codes from <https://ethervm.io/>
+      if (param === undefined) {
+        const opcode = OPCODES[op];
+        if (opcode === undefined) {
+          throw new Error(`unsupported op ${op}`);
+        }
+        return [opcode];
+      } else {
+        if (op.toUpperCase() !== "PUSH") {
+          throw new Error(`unsupported parameterized op ${op}`);
+        }
+
+        // compute the byte-encoded value for this parameter. Note that we use
+        // the ABI coder to ensure it fits into a `uint256` (`BigNumber` has
+        // arbitrary size) and use `toHexString` to strip leading 0's with
+        // special handling for the 0 value.
+        const value = Array.from(
+          ethers.utils.arrayify(
+            BigNumber.from(
+              ethers.utils.defaultAbiCoder.encode(["uint256"], [param]),
+            ).toHexString(),
+          ),
+        );
+        assert(value.length > 0 && value.length <= 32);
+
+        const opcode = PUSH1_OPCODE + value.length - 1;
+        return [opcode, ...value];
+      }
+    })
+    .reduce((bytecode, opcodes) => [...bytecode, ...opcodes], []);
```

Comment:
Could this be replaced with `.flat()`?

Reply:
I had `flatMap` originally, but we build for ES6 as well for the FE which doesn't support `flat` or `flatMap` (at least that is what the TS error suggests).